From 3c70e5ea62158b5a09445065ab362ddbfed4990c Mon Sep 17 00:00:00 2001
From: Taso Dane <taso@google.com>
Date: Fri, 1 May 2026 02:40:50 -0700
Subject: [PATCH 18/49] Strip URI grants in ConfigActivityProxy

AppWidgetConfigActivityProxy is a system-privileged activity that
relays the result of a widget configuration activity back to the
host. Since the proxy's PendingIntent is created by the system,
calling setResult() with an intent containing URI permission grant
flags would cause the system to grant those permissions to the
caller. This allows a malicious widget provider to trick the system
into granting it unauthorized URI permissions.

This change strips the URI permission grant flags from the result
intent before calling setResult() in the proxy activity. This
ensures that no URI permissions are inadvertently granted by the
system UID.

Bug: 498916138
Test: atest FrameworksCoreTests:android.appwidget.AppWidgetConfigActivityProxyTest
Flag: EXEMPT BUGFIX

Autogenerated by MendIt (go/androidmendit)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:e2c75d70260ad97f24feef6b8ec1a0ebfe364347
Merged-In: I770b26d0e28e1090abae452e3bc00af51bf9b30e
Change-Id: I770b26d0e28e1090abae452e3bc00af51bf9b30e
---
 .../AppWidgetConfigActivityProxy.java         |   9 +-
 core/tests/coretests/Android.bp               |  11 ++
 core/tests/coretests/AndroidManifest.xml      |   6 +
 .../AppWidgetConfigActivityProxyTest.kt       | 109 ++++++++++++++++++
 .../src/android/appwidget/TEST_MAPPING        |   7 ++
 5 files changed, 137 insertions(+), 5 deletions(-)
 create mode 100644 core/tests/coretests/src/android/appwidget/AppWidgetConfigActivityProxyTest.kt
 create mode 100644 core/tests/coretests/src/android/appwidget/TEST_MAPPING

diff --git a/core/java/android/appwidget/AppWidgetConfigActivityProxy.java b/core/java/android/appwidget/AppWidgetConfigActivityProxy.java
index d6a87d3d653a..d6874127c7b5 100644
--- a/core/java/android/appwidget/AppWidgetConfigActivityProxy.java
+++ b/core/java/android/appwidget/AppWidgetConfigActivityProxy.java
@@ -21,10 +21,6 @@ import android.content.ActivityNotFoundException;
 import android.content.Intent;
 import android.os.Bundle;
 import android.util.Log;
-import android.view.View;
-import android.view.ViewGroup;
-import android.view.WindowManager;
-import android.view.WindowManagerWrapper;
 
 import androidx.annotation.Nullable;
 
@@ -64,7 +60,10 @@ public class AppWidgetConfigActivityProxy extends Activity {
 
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
-        setResult(resultCode, data);
+        // b/498916138 - Only pass the result code and intent extras to the config activity and
+        // nothing else.
+        Intent result = data != null ? new Intent().putExtras(data) : null;
+        setResult(resultCode, result);
         int widgetId = getIntent().getIntExtra(
                 AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
         AppWidgetManager.getInstance(this).setConfigActivityComplete(widgetId);
diff --git a/core/tests/coretests/Android.bp b/core/tests/coretests/Android.bp
index 4573386a6b03..7d61bf5946f2 100644
--- a/core/tests/coretests/Android.bp
+++ b/core/tests/coretests/Android.bp
@@ -591,6 +591,17 @@ test_module_config {
     include_filters: ["com.android.internal.util.LatencyTrackerTest"],
 }
 
+test_module_config {
+    name: "FrameworksCoreTests_android_appwidget",
+    base: "FrameworksCoreTests",
+    test_suites: [
+        "automotive-tests",
+        "device-platinum-tests",
+        "device-tests",
+    ],
+    include_filters: ["android.appwidget."],
+}
+
 test_module_config {
     name: "FrameworksCoreTests_content_capture_options",
     base: "FrameworksCoreTests",
diff --git a/core/tests/coretests/AndroidManifest.xml b/core/tests/coretests/AndroidManifest.xml
index a786166eaec3..3dfffdd3a933 100644
--- a/core/tests/coretests/AndroidManifest.xml
+++ b/core/tests/coretests/AndroidManifest.xml
@@ -187,6 +187,12 @@
         <meta-data
             android:name="android.content.APP_RESTRICTIONS"
             android:resource="@xml/app_restrictions" />
+        <activity android:name="android.appwidget.FakeConfigActivity"
+                  android:exported="true" />
+        <activity android:name="android.appwidget.FakeConfigWithFlagsActivity"
+                  android:exported="true" />
+        <activity android:name="android.appwidget.AppWidgetConfigActivityProxy"
+                  android:exported="true" />
         <activity android:name="android.view.ViewAttachTestActivity"
             android:label="View Attach Test"
             android:exported="true">
diff --git a/core/tests/coretests/src/android/appwidget/AppWidgetConfigActivityProxyTest.kt b/core/tests/coretests/src/android/appwidget/AppWidgetConfigActivityProxyTest.kt
new file mode 100644
index 000000000000..4562b301693a
--- /dev/null
+++ b/core/tests/coretests/src/android/appwidget/AppWidgetConfigActivityProxyTest.kt
@@ -0,0 +1,109 @@
+/*
+ * Copyright (C) 2026 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package android.appwidget
+
+import android.app.Activity
+import android.content.ComponentName
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+import androidx.test.core.app.ActivityScenario.launchActivityForResult
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+
+class FakeConfigWithFlagsActivity : Activity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val data = Intent().apply {
+            setDataAndType(Uri.parse("content://"), "image/*")
+            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
+            addFlags(Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
+            addFlags(Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION)
+            addFlags(Intent.FLAG_GRANT_PREFIX_URI_PERMISSION)
+        }
+        setResult(RESULT_OK, data)
+        finish()
+    }
+}
+
+class FakeConfigActivity : Activity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        val data = Intent()
+        data.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, WIDGET_ID)
+        setResult(RESULT_OK, data)
+        finish()
+    }
+
+    companion object {
+        const val WIDGET_ID = 123
+    }
+}
+
+@RunWith(AndroidJUnit4::class)
+class AppWidgetConfigActivityProxyTest {
+
+    @Test
+    fun onActivityResult_preservesAppWidgetIdExtra() {
+        val intent = Intent(Intent.ACTION_MAIN)
+        intent.component = PROXY_COMPONENT
+
+        val targetIntent = Intent()
+        targetIntent.component = ComponentName(
+            "com.android.frameworks.coretests",
+            FakeConfigActivity::class.java.name
+        )
+        intent.putExtra(Intent.EXTRA_INTENT, targetIntent)
+
+        launchActivityForResult<AppWidgetConfigActivityProxy>(intent).use { scenario ->
+            assertThat(scenario.result.resultCode).isEqualTo(Activity.RESULT_OK)
+            val resultData = scenario.result.resultData
+            assertThat(resultData.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1))
+                .isEqualTo(FakeConfigActivity.WIDGET_ID)
+        }
+    }
+
+    @Test
+    fun onActivityResult_stripsUriPermissions() {
+        val intent = Intent(Intent.ACTION_MAIN)
+        intent.component = PROXY_COMPONENT
+
+        val targetIntent = Intent()
+        targetIntent.component = ComponentName(
+            "com.android.frameworks.coretests",
+            FakeConfigWithFlagsActivity::class.java.name
+        )
+        intent.putExtra(Intent.EXTRA_INTENT, targetIntent)
+
+        launchActivityForResult<AppWidgetConfigActivityProxy>(intent).use { scenario ->
+            assertThat(scenario.result.resultCode).isEqualTo(Activity.RESULT_OK)
+            val resultData = scenario.result.resultData
+            assertThat(resultData.flags).isEqualTo(0)
+            assertThat(resultData.data).isNull()
+        }
+    }
+
+    companion object {
+        private val PROXY_COMPONENT =
+            ComponentName(
+                "com.android.frameworks.coretests",
+                AppWidgetConfigActivityProxy::class.java.name
+            )
+    }
+}
diff --git a/core/tests/coretests/src/android/appwidget/TEST_MAPPING b/core/tests/coretests/src/android/appwidget/TEST_MAPPING
new file mode 100644
index 000000000000..2a2a29c3fc22
--- /dev/null
+++ b/core/tests/coretests/src/android/appwidget/TEST_MAPPING
@@ -0,0 +1,7 @@
+{
+  "presubmit": [
+    {
+      "name": "FrameworksCoreTests_android_appwidget"
+    }
+  ]
+}
-- 
2.55.0

